include/libpmemobj++/experimental/vector.hpp

```diff
@@ -0,0 +1,399 @@
+	void _construct_at_end(size_type count, const value_type &value);
+	template <typename InputIt,
+		  typename std::enable_if<
+			  detail::is_input_iterator<InputIt>::value &&
+				  std::is_constructible<
+					  value_type,
+					  typename std::iterator_traits<
+						  InputIt>::reference>::value,
+			  InputIt>::type * = nullptr>
+	void _construct_at_end(InputIt first, InputIt last);
+	void _destruct_at_end(size_type size_new) noexcept;
+
+	/* Underlying array */
+	persistent_ptr<T[]> _data = nullptr;
+
+	p<size_type> _size = 0;
+	p<size_type> _capacity = 0;
+};
+
+/**
+ * Default constructor. Constructs an empty container.
+ *
+ * @throw pmem::pool_error if an object is not in persistent memory.
+ */
+template <typename T>
+vector<T>::vector() : _data(nullptr), _size(0), _capacity(0)
+{
+	if (nullptr == pmemobj_pool_by_ptr(this))
+		throw pool_error("Invalid pool handle.");
+}
+
+/**
+ * Checks whether the container is empty.
+ *
+ * @return true if container is empty, 0 othwerwise.
+ */
+template <typename T>
+constexpr bool
+vector<T>::empty() const noexcept
+{
+	return _size == 0;
+}
+
+/**
+ * @return maximum number of elements the container is able to hold due to PMDK
+ * limitations.
+ */
+template <typename T>
+constexpr typename vector<T>::size_type
+vector<T>::max_size() const noexcept
+{
+	return PMEMOBJ_MAX_ALLOC_SIZE / sizeof(value_type);
+}
+
+/**
+ * Private helper function. Must be called during transaction. Allocates memory
+ * for given number of elements.
+ *
+ * @param[in] capacity_new capacity of new underlying array
+ *
+ * @throw std::length_error if new size exceeds biggest possible pmem
+ * allocation.
+ * @throw pmem::transaction_alloc_error when allocating memory for underlying
+ * array in transaction failed.
+ *
+ * @pre pmemobj_tx_stage() == TX_STAGE_WORK
+ *
+ * @post _capacity = capacity_new
+ */
+template <typename T>
+void
+vector<T>::_alloc(size_type capacity_new)
+{
+	assert(pmemobj_tx_stage() == TX_STAGE_WORK);
+
+	_capacity = capacity_new;
+
+	if (capacity_new == 0) {
+		_data = nullptr;
+		return;
+	}
+
+	if (capacity_new > max_size())
+		throw std::length_error("New capacity exceeds max size.");
+
+	_data = pmemobj_tx_alloc(sizeof(value_type) * capacity_new,
+				 detail::type_num<value_type>());
+
+	if (_data == nullptr)
+		throw transaction_alloc_error(
+			"failed to allocate persistent memory object");
+}
+
+/**
+ * Private helper function. Must be called during transaction. Dellocates
+ * underlying array.
+ *
+ * @throw pmem::transaction_free_error when freeing old underlying array failed.
+ *
+ * @pre pmemobj_tx_stage() == TX_STAGE_WORK
+ *
+ * @post _size = _capacity = 0
+ * @post _data = nullptr
+ */
+template <typename T>
+void
+vector<T>::_dealloc()
+{
+	assert(pmemobj_tx_stage() == TX_STAGE_WORK);
+
+	if (_data != nullptr) {
+		_destruct_at_end(0);
+		if (pmemobj_tx_free(*_data.raw_ptr()) != 0)
+			throw transaction_free_error(
+				"failed to delete persistent memory "
+				"object");
```

Comment:
Please add "BreakStringLiterals: false" to .clang-format first and then unbreak this line.